weavertest/multi.go

```diff
@@ -86,19 +90,81 @@ func initMultiProcess(ctx context.Context, t testing.TB, config string) weaver.I
 		t.Fatal(err)
 	}
 
-	// Deploy main.
-	toWeavelet, toEnvelope, err := b.CreateAndRunEnvelopeForMain(createWeaveletForMain(dep), dep.App)
+	// Set up the pipes between the envelope and the main weavelet. The
+	// pipes will be closed by the envelope and weavelet conns.
+	//
+	//         envelope                      weavelet
+	//         --------        +----+        --------
+	//   fromWeaveletReader <--| OS |<-- fromWeaveletWriter
+	//     toWeaveletWriter -->|    |--> toWeaveletReader
+	//                         +----+
+	fromWeaveletReader, fromWeaveletWriter, err := os.Pipe()
 	if err != nil {
-		t.Fatal(err)
+		t.Fatal(fmt.Errorf("cannot create fromWeavelet pipe: %v", err))
+	}
+	toWeaveletReader, toWeaveletWriter, err := os.Pipe()
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create toWeavelet pipe: %v", err))
+	}
+
+	weaveletInfo := createWeaveletForMain(dep)
+
+	// Create a connection between the weavelet for the main process and the envelope.
+	econn, err := conn.NewEnvelopeConn(fromWeaveletReader, toWeaveletWriter, b, weaveletInfo)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope conn: %v", err))
 	}
+
+	// Create and run an envelope for the main process.
+	options := envelope.Options{
+		Restart:         envelope.Never,
+		Retry:           retry.DefaultOptions,
+		GetEnvelopeConn: func() *conn.EnvelopeConn { return econn },
+	}
+	e, err := envelope.NewEnvelope(weaveletInfo, dep.App, b, options)
+	if err != nil {
+		t.Fatal(fmt.Errorf("cannot create envelope: %v", err))
+	}
+
+	go econn.Run()
+	go e.Run(ctx)
```

Comment:
Now that this logic is in a function with `t.Cleanup`, these errors could be  handled by starting the goroutines with https://pkg.go.dev/golang.org/x/sync/errgroup.

The `t.Cleanup` function can call `group.Wait` and receive the errors. More about why that doesn't work yet below.

Another option would be to simply send the error to a channel, and check that the channel is empty in `t.Cleanup`. That avoids the need to wait on the goroutine exit.

Reply:
After posting the initial change I realized there is a 3rd option that is much easier. I can use `t.Error` here to capture the error now that this has access to `testing.T`. I've made that change in the latest commit and marked the PR ready for review.